File: typescript/src/sprites/particle.ts
import P5 from 'p5';
import Colors from '../ui/colors';
import Sprite from './sprite';

const POLYGON_SIDES = 5;
const HEALTH_DECREMENT = 5;
const MINIMAL_HEALTH = 50;

export default class Particle extends Sprite {
    private health: number;

    constructor(p5: P5, position: P5.Vector, diameter: number) {
        super(p5, position, diameter, P5.Vector.random2D());

        this.health = p5.random(175, 230);
    }

    update(): boolean {
        super.update();

        this.health -= HEALTH_DECREMENT;

        if (this.health <= MINIMAL_HEALTH) {
            return false;
        }

        return true;
    }

    draw() {
        this.p5.push();

        this.p5.stroke(Colors.EDGE);
        this.p5.noFill();

        this.polygon(this.position, this.diameter, POLYGON_SIDES);

        this.p5.pop();
    }

    private polygon(position: P5.Vector, radius: number, sides: number) {
        let angle = this.p5.TWO_PI / sides;

        this.p5.beginShape();

        for (let a = 0; a < this.p5.TWO_PI; a += angle) {
            let sx = position.x + this.p5.cos(a) * radius;
            let sy = position.y + this.p5.sin(a) * radius;

            this.p5.vertex(sx, sy);
        }

        this.p5.endShape(this.p5.CLOSE);
    }
}

File: rust/src/sprites/particle.rs
use std::f64::consts::PI;

use web_sys::CanvasRenderingContext2d;

use crate::utils::{config::Config, general::random, vector::Vector};

use super::sprite::{CanvasDimension, Spritable, Sprite, SpriteData};

pub struct Particle {
    health: u32,
    pub is_faded: bool,
    pub sprite: Sprite,
    config: Config,
}

impl Spritable for Particle {
    fn update(&mut self) {
        self.sprite.update();

        if self.health < self.config.sprites.particle.minimal_health {
            self.is_faded = true;
        } else {
            self.health -= self.config.sprites.particle.health_decrement;
        }
    }

    fn draw(&self, canvas: CanvasRenderingContext2d) {
        if !self.is_faded {
            canvas.save();

            self.polygon(
                self.sprite.data.position,
                self.sprite.data.diameter,
                self.config.sprites.particle.polygon_sides,
                canvas.clone(),
            );

            canvas.restore();
        }
    }

    fn collide_with(&self, sprite: Sprite) -> bool {
        self.sprite.collide_with(sprite)
    }
}

impl Particle {
    pub fn new(sprite_data: SpriteData, canvas: CanvasDimension) -> Particle {
        let config = Config::new();

        let mut new_sprite_data = sprite_data;
        new_sprite_data.diameter = random(1, 5) as f64;
        new_sprite_data.velocity = Vector::random(-1.2, 1.2);

        Particle {
            health: random(175, 230),
            is_faded: false,
            sprite: Sprite::new(new_sprite_data, canvas),
            config,
        }
    }

    fn polygon(&self, position: Vector, radius: f64, sides: u8, canvas: CanvasRenderingContext2d) {
        let angle = 2.0 * PI / sides as f64;

        canvas.begin_path();

        canvas.move_to(position.x + radius, position.y);

        for side in 1..sides {
            let current_angle = angle * side as f64;
            let sx = position.x + current_angle.cos() * radius;
            let sy = position.y + current_angle.sin() * radius;

            canvas.line_to(sx, sy);
        }

        canvas.close_path();

        canvas.stroke();
    }
}
